filters/builtin/decompress.go

```diff
@@ -0,0 +1,231 @@
+)
+
+const (
+	// DecompressionNotPossible is the state-bag key to indicate
+	// to the subsequent filters during response processing that the
+	// content is compressed, but decompression was not possible, e.g
+	// because the encoding is not supported.
+	DecompressionNotPossible = "filter::decompress::impossible"
+
+	// DecompressionError is the state-bag key to indicate to the
+	// subsequent filters during response processing that th
+	// decompression of the content was attempted but failed. The
+	// response body may have been sniffed, and therefore it was
+	// discarded.
+	DecompressionError = "filter::decompress::error"
+)
+
+type decodedBody struct {
+	enc        string
+	original   io.Closer
+	decoder    io.ReadCloser
+	isFromPool bool
+}
+
+type decodingError struct {
+	decoder  error
+	original error
+}
+
+type decompress struct{}
+
+var supportedEncodingsDecompress = map[string]*sync.Pool{
+	"gzip":    {},
+	"deflate": {},
+}
+
+func init() {
+	// #cpu * 4: pool size decided based on some
+	// simple tests, checking performance by binary
+	// steps (https://github.com/zalando/skipper)
+	for enc, pool := range supportedEncodingsDecompress {
+		for i := 0; i < runtime.NumCPU()*4; i++ {
+			pool.Put(newDecoder(enc))
+		}
+	}
+}
+
+func newDecoder(enc string) io.ReadCloser {
+	switch enc {
+	case "gzip":
+		return new(gzip.Reader)
+	default:
+		return flate.NewReader(nil)
+	}
+}
+
+func fromPool(enc string) (io.ReadCloser, bool) {
+	d, ok := supportedEncodingsDecompress[enc].Get().(io.ReadCloser)
+	return d, ok
+}
+
+func reset(decoder, original io.ReadCloser, enc string) error {
+	switch enc {
+	case "gzip":
+		return decoder.(*gzip.Reader).Reset(original)
+	default:
+		return decoder.(flate.Resetter).Reset(original, nil)
+	}
+}
+
+func newDecodedBody(original io.ReadCloser, encs []string) (body io.ReadCloser, err error) {
+	if len(encs) == 0 {
+		body = original
+		return
+	}
+
+	last := len(encs) - 1
+	enc := encs[last]
+	encs = encs[:last]
+
+	decoder, isFromPool := fromPool(enc)
+	if !isFromPool {
+		decoder = newDecoder(enc)
+	}
+
+	if err = reset(decoder, original, enc); err != nil {
+		return
+	}
+
+	decoded := decodedBody{
+		enc:        enc,
+		original:   original,
+		decoder:    decoder,
+		isFromPool: isFromPool,
+	}
+
+	return newDecodedBody(decoded, encs)
+}
+
+func (b decodedBody) Read(p []byte) (int, error) {
+	return b.decoder.Read(p)
+}
+
+func (b decodedBody) Close() error {
+	derr := b.decoder.Close()
+	if b.isFromPool {
+		if derr != nil {
+			supportedEncodingsDecompress[b.enc].Put(newDecoder(b.enc))
+		} else {
+			supportedEncodingsDecompress[b.enc].Put(b.decoder)
+		}
+	}
+
+	oerr := b.original.Close()
+	var err error
+	if derr != nil || oerr != nil {
+		err = decodingError{
+			decoder:  derr,
+			original: oerr,
+		}
+	}
+
+	return err
+}
+
+func (e decodingError) Error() string {
+	switch {
+	case e.decoder == nil:
+		return e.original.Error()
+	case e.original == nil:
+		return e.decoder.Error()
```

Comment:
I think it would be less confusing if you used != nil, like this:
```
case e.decoder != nil:
		return e.decoder.Error()
	case e.original != nil:
		return e.original.Error()
```